content/en/blog/_posts/2022-10-20-advanced-server-side-apply.md

```diff
@@ -0,0 +1,81 @@
+
+#### CI/CD systems {#ci-cd-systems}
+Continuous integration (CI) and/or continuous deployment (CD) systems are a special kind of controller which is doing something like reading manifests from source control (such as a Git repo) and automatically pushing them into the cluster. Perhaps the CI / CD process first generates manifests from a template, then runs some tests, and then deploys a change. Typically, users are the entities pushing changes into source control, although that’s not necessarily always the case.
+
+Some systems like this continuously reconcile with the cluster, others may only operate when a change is pushed to the source control system. The following considerations are important for both, but more so for the continuously reconciling kind.
+
+CI/CD systems are literally controllers, but for the purpose of **apply**, they are more like users, and unlike other controllers, they need to pay attention to conflicts. Reasoning:
+* Abstractly, CI/CD systems can change anything, which means they could conflict with any controller out there. The recommendation that controllers force conflicts is assuming that controllers change a limited number of things and you can be reasonably sure that they won’t fight with other controllers about those things; that’s clearly not the case for CI/CD controllers.
+* Concrete example: imagine the CI/CD system wants `.spec.replicas` for some Deployment to be 3, because that is the value that is checked into source code; however there is also a HorizontalPodAutoscaler (HPA) that targets the same deployment. The HPA computes a target scale and decides that there should be 10 replicas. Which should win? I just said that most controllers–including the HPA–should ignore conflicts. The HPA has no idea if it has been enabled incorrectly, and the HPA has no convenient way of informing users of errors.
+* The other common cause of a CI/CD system getting a conflict is probably when it is trying to overwrite a hot-fix (hand-rolled patch) placed there by a system admin / SRE / dev-on-call. You almost certainly don’t want to override that automatically.
+* Of course, sometimes SRE makes an accidental change, or a dev makes an unauthorized change – those you do want to notice and overwrite; however, the CI/CD system can’t tell the difference between these last two cases.
+
+Hopefully this convinces you that CI/CD systems need error paths–a way to back-propagate these conflict errors to humans; in fact, they should have this already, certainly continuous integration systems need some way to report that tests are failing. But maybe we can say something about how humans can deal with errors:
+* Reject the hotfix: the system admin observes the error, and manually force-applies the manifest in question. Then the CI/CD system will be able to apply the manifest successfully and become a co-owner. (Optional: then the system admin applies a blank manifest to relinquish any fields they became a manager for.)
```

Comment:
I wrote "system admin" because often (or at least possibly) the developer doesn't have the permissions the CI/CD system has and literally cannot do this. I can improve this wording in other ways.

Reply:
hopefully it is clearer now.